CONTRIBUTING.md

```diff
@@ -121,3 +122,115 @@ run it before submitting your PR. It runs `gofmt -s` (simplify) and `golint`.
+are important for software that runs autonomously and continuously without direct
+interaction with a human (such as this Collector).
+
+### Startup Error Handling
+
+Verify configuration during startup and fail fast if the configuration is invalid.
+This will bring attention of a human to the problem as it is more typical for humans
+to notice problems when the process is starting as opposed to problems that may arise
+sometime (potentially long time) after process startup. Monitoring systems are likely
+to automatically flag processes that exit with failure during startup, making it
+easier to notice the problem. The Collector should print a reasonable log message to
+explain the problem and exit with a non-zero code. It is acceptable to crash the process
+during startup if there is no good way to exit cleanly but do your best to log and
+exit cleanly with a process exit code.
+
+### Do not Crash after Startup
+
+Do not crash or exit the Collector process after the startup sequence is finished.
+A running Collector typically contains data that is received but not yet exported further
+(e.g. is stored in the queues and other processors). Crashing or exiting Collector
+process will result in losing this data since typically the receiver have
+already acknowledged the receipt for this data and the senders of the data will
+not send that data again. This recommendation may change in the future if we
+implement persistent queues in the Collector.
+
+### Bad Input Handling
+
+Do not crash on bad input in receivers or elsewhere in the pipeline.
+[Crash-only software](https://en.wikipedia.org/wiki/Crash-only_software)
+is valid in certain cases however this is not a correct approach for Collector (except
+during startup, see above). The reason is that many senders from which Collector
+receives data have built-in automatic retries of the _same_ data if no
+acknowledgment is received from the Collector. If you crash on bad input
+chances are high that after the Collector is restarted it will see the same
+data in the input and will crash again. This will likely result in infinite
+crashing loop if you have automatic retries in place.
+
+### Error Handling and Retries
+
+Be rigorous in error handling. Don't ignore errors. Think carefully about each
+error and decide if it is a fatal problem or a transient problem that may go away
+when retried. For fatal errors that the users of Collector need to be aware of
+log them or have an internal metric that counts occurrences. For transient
+errors come up with a retrying strategy and implement it. Typically you will
+want to implement retries with some sort of exponential back-off strategy. For
+connection or sending retries use jitter for back-off intervals to avoid overwhelming
+your destination when network is restored or the destination is recovered.
+[Exponential Backoff](https://github.com/cenkalti/backoff) is a good library that
+provides all this functionality.
+
+### Logging
```

Comment:
Are there a specific set of log levels devs should be using?  Should that be mentioned here?

Reply:
Added a paragraph on log levels.

Reply:
I re-read what I wrote and don't quite like it, too verbose. Removing for now. Should be added later.